### ISSUE TITLE
missing tautological-compare warning for float == double constant 

### ISSUE THREAD

**Issue by zygoloid** [llvm/llvm-project#54222]
Example:
```c++
float f = 0.1;
if (f == 0.1) { ... }
```

This is a tautological comparison: the LHS of the comparison is always a `float` value promoted to `double`, and the RHS of the comparison is a value that cannot result from a `float`-to-`double` promotion.

We should probably also warn on `f != non_float_constant`; such comparisons are not tautological, but are equivalent to `!isnan(f)`. We shouldn't warn if the constant value is exactly representable in type `float`. And don't warn on `<`, `<=`, `>=`, `>`, `<=>`.

**Comment by rotateright** [comment]
We also miss this (and related) folds in the IR optimizer.

**Comment by rotateright** [comment]
Adding the optimizer before the warning was probably a mistake. :)
Here's a proposal for the first part of this request:
https://reviews.llvm.org/D121306

**Comment by rotateright** [comment]
> We should probably also warn on `f != non_float_constant`; such comparisons are not tautological, but are equivalent to `!isnan(f)`.

"(float f) != 0.1"
is tautological (always true) as discussed in D121306, and we should warn on that too with ab982eace6.

So I think we can close this report, but please reopen if I missed something. I'll plan to recommit the optimizer change next week.
